controllers/ssp_controller.go

```diff
@@ -84,67 +85,227 @@ func (r *SSPReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
+}
 
-	return ctrl.Result{}, nil
+func (r *SSPReconciler) clearCache() {
+	r.LastSspSpec = ssp.SSPSpec{}
+	r.SubresourceCache = common.VersionCache{}
 }
 
 func isBeingDeleted(object metav1.Object) bool {
 	return !object.GetDeletionTimestamp().IsZero()
 }
 
-func initialize(request *common.Request) (bool, error) {
-	changed := false
+func isInitialized(ssp *ssp.SSP) bool {
+	return isBeingDeleted(ssp) || ssp.Status.Phase != lifecycleapi.PhaseEmpty
+}
 
-	if !isBeingDeleted(request.Instance) {
-		if !controllerutil.ContainsFinalizer(request.Instance, finalizerName) {
-			controllerutil.AddFinalizer(request.Instance, finalizerName)
-			changed = true
-		}
+func initialize(request *common.Request) error {
+	controllerutil.AddFinalizer(request.Instance, finalizerName)
+	err := request.Client.Update(request.Context, request.Instance)
+	if err != nil {
+		return err
 	}
 
-	var err error
-	if changed {
-		err = request.Client.Update(request.Context, request.Instance)
-	}
-	return changed, err
+	request.Instance.Status.Phase = lifecycleapi.PhaseDeploying
+	return request.Client.Status().Update(request.Context, request.Instance)
 }
 
 func cleanup(request *common.Request) error {
-	if !controllerutil.ContainsFinalizer(request.Instance, finalizerName) {
+	if controllerutil.ContainsFinalizer(request.Instance, finalizerName) {
+		request.Instance.Status.Phase = lifecycleapi.PhaseDeleting
+		err := request.Client.Status().Update(request.Context, request.Instance)
+		if err != nil {
+			return err
+		}
+		for _, operand := range sspOperands {
+			err = operand.Cleanup(request)
+			if err != nil {
+				return err
+			}
+		}
+		controllerutil.RemoveFinalizer(request.Instance, finalizerName)
+		err = request.Client.Update(request.Context, request.Instance)
+		if err != nil {
+			return err
+		}
+	}
+
+	request.Instance.Status.Phase = lifecycleapi.PhaseDeleted
+	err := request.Client.Status().Update(request.Context, request.Instance)
+	if errors.IsConflict(err) || errors.IsNotFound(err) {
+		// These errors are ignored. They can happen if the CR was removed
+		// before the status update call is executed.
 		return nil
 	}
+	return err
+}
 
+func reconcileOperands(sspRequest *common.Request) ([]common.ResourceStatus, error) {
+	allStatuses := make([]common.ResourceStatus, 0, len(sspOperands))
 	for _, operand := range sspOperands {
-		err := operand.Cleanup(request)
+		statuses, err := operand.Reconcile(sspRequest)
 		if err != nil {
-			return err
+			return nil, err
+		}
+		allStatuses = append(allStatuses, statuses...)
+	}
+	return allStatuses, nil
+}
+
+func updateStatus(request *common.Request, statuses []common.ResourceStatus) error {
+	notAvailable := make([]common.ResourceStatus, 0, len(statuses))
+	progressing := make([]common.ResourceStatus, 0, len(statuses))
+	degraded := make([]common.ResourceStatus, 0, len(statuses))
+	for _, status := range statuses {
+		if status.NotAvailable != nil {
+			notAvailable = append(notAvailable, status)
```

Comment:
Are `Available`, `Progressing` and `Degraded` mutually exclusive?
If not, what happens in the case of 2 `Progressing` resources, is `Available` being reported on the CR because `len(notAvailable) = 0`?

Reply:
They are related but not mutually exclusive. These are the meanings:
- `Available` - the resource is available to perform the expected task. For example a Deployment that has at least 1 pod running.
- `Progressing` - the resource has not finished deploying, but it can be available. For example a Deployment where some pods are still starting.
- `Degraded` - the resource is not completely deployed. This can mean that it is still `progressing`, but it can also mean that there was an error during deployment.

Currently on SSP CR, `progressing` and `degraded` always have the same value.

Reply:
Ah, I did not know that `Available` was reported when a Deployment was not fully rolled out, that's good to know.
So, a situation with all `Available`, `Degraded`, `Progressing` is a valid one?

Reply:
That is how I understood it, so I implemented it that way. A situation when all these conditions are true happens if some pods for `template-validator` are running, but not all of them.